pkg/vm/engine/tae/catalog/base.go

```diff
@@ -22,466 +22,569 @@ import (
+	if un == nil {
+		return false
 	}
-	return cloned
+	if un.IsActive() {
+		return false
+	}
+	return un.HasDropped()
+}
+
+func (be *BaseEntry) PrepareWrite(txn txnif.TxnReader, rwlocker *sync.RWMutex) (err error) {
+	node := be.GetUpdateNodeLocked()
+	if node.IsActive() {
+		if node.IsSameTxn(txn.GetStartTS()) {
+			return
+		}
+		return txnif.ErrTxnWWConflict
+	}
+
+	if node.Start.Greater(txn.GetStartTS()) {
+		return txnif.ErrTxnWWConflict
+	}
+	return
 }
 
-func (be *BaseEntry) WriteTo(w io.Writer) (n int64, err error) {
+func (be *BaseEntry) WriteOneNodeTo(w io.Writer) (n int64, err error) {
 	if err = binary.Write(w, binary.BigEndian, be.ID); err != nil {
 		return
 	}
-	if err = binary.Write(w, binary.BigEndian, be.CreateAt); err != nil {
+	n += 8
+	var n2 int64
+	n2, err = be.GetUpdateNodeLocked().WriteTo(w)
+	if err != nil {
 		return
 	}
-	if err = binary.Write(w, binary.BigEndian, be.DeleteAt); err != nil {
+	n += n2
+	return
+}
+
+func (be *BaseEntry) WriteAllTo(w io.Writer) (n int64, err error) {
+	if err = binary.Write(w, binary.BigEndian, be.ID); err != nil {
 		return
 	}
-	sn := int64(0)
-	if sn, err = be.CommitInfo.WriteTo(w); err != nil {
+	n += 8
+	if err = binary.Write(w, binary.BigEndian, be.length); err != nil {
 		return
 	}
-	n = sn + 8 + 8 + 8
+	n += 8
+	be.MVCC.Loop(func(node *common.DLNode) bool {
+		var n2 int64
+		n2, err = node.GetPayload().(*UpdateNode).WriteTo(w)
+		if err != nil {
+			return false
+		}
+		n += n2
+		return true
+	}, true)
 	return
 }
 
-func (be *BaseEntry) ReadFrom(r io.Reader) (n int64, err error) {
+func (be *BaseEntry) ReadOneNodeFrom(r io.Reader) (n int64, err error) {
 	if err = binary.Read(r, binary.BigEndian, &be.ID); err != nil {
 		return
 	}
-	if err = binary.Read(r, binary.BigEndian, &be.CreateAt); err != nil {
+	var n2 int64
+	un := NewEmptyUpdateNode()
+	n2, err = un.ReadFrom(r)
+	if err != nil {
 		return
 	}
-	if err = binary.Read(r, binary.BigEndian, &be.DeleteAt); err != nil {
-		return
-	}
-	sn := int64(0)
-	sn, err = be.CommitInfo.ReadFrom(r)
-	n = sn + 8 + 8 + 8
-	be.RWMutex = new(sync.RWMutex)
+	be.InsertNode(un)
+	n += n2
 	return
 }
 
-func (be *BaseEntry) GetTxn() txnif.TxnReader { return be.Txn }
-
-func (be *BaseEntry) IsTerminated(waitIfcommitting bool) bool {
-	return be.Txn.IsTerminated(waitIfcommitting)
-}
-
-func (be *BaseEntry) IsCommitted() bool {
-	return be.Txn == nil && !be.CreateAt.IsEmpty()
+func (be *BaseEntry) ReadAllFrom(r io.Reader) (n int64, err error) {
+	if err = binary.Read(r, binary.BigEndian, &be.ID); err != nil {
+		return
+	}
+	n += 8
+	if err = binary.Read(r, binary.BigEndian, &be.length); err != nil {
+		return
+	}
+	n += 8
+	for i := 0; i < int(be.length); i++ {
+		var n2 int64
+		un := NewEmptyUpdateNode()
+		n2, err = un.ReadFrom(r)
+		if err != nil {
+			return
+		}
+		be.MVCC.Insert(un)
+		n += n2
+	}
+	return
 }
 
-func (be *BaseEntry) GetID() uint64 { return be.ID }
-
 func (be *BaseEntry) DoCompre(oe *BaseEntry) int {
```

Comment:
Now you introduced an compare on be.id and oe.id -- does this ordering based on id, actually always work?  I don't know, but ask Xu Peng to confirm.

Will the id of be and oe ever change?      If not, why this rlock/runlock is useful?